src/sqlfluff/core/parser/segments/base.py

```diff
@@ -632,19 +665,24 @@ def is_optional(cls):
     @classmethod
     def class_is_type(cls, *seg_type):
         """Is this segment class (or its parent) of the given type."""
-        # Do we match on the type of _this_ class.
-        if cls.type in seg_type:
-            return True
-        # If not, check types of parents.
-        for base_class in cls.__bases__:
-            if base_class is object:
-                break
-            elif base_class.type in seg_type:
+        # Work backward in the class heirachy to see if there's a match
+        for class_type in cls._class_types():
+            if class_type in seg_type:
                 return True
-            elif base_class.type == "base":
-                break
         return False
 
+    @classmethod
+    def _class_types(cls) -> Iterator[str]:
```

Comment:
Think it'd be worth precomputing this at class definition time? (Potential performance improvement.)

Reply:
That's a great idea.... but how? I'm not aware of how to compute and store class attribute on definition. @barrywhart are you?

Best I can think of would be another cached property - but that's not computed on definition, just on first call.

Reply:
Metaclasses are great for this. They basically provide a way to call a custom hook on `BaseSegment` that is called at class definition time (for `BaseSegment` or any subclass of it).

I threw this together quickly, but I think it will do what we need. It's adapted from the example in this [tutorial](https://www.pythontutorial.net/python-oop/python-metaclass-example/).

```
class SegmentMetaclass(type):
    def __new__(mcs, name, bases, class_dict):
        class_obj = super().__new__(mcs, name, bases, class_dict)
        class_types = {class_dict["type"]}
        for base in bases:
            class_types.update(base.class_types)
        class_obj.class_types = class_types
        return class_obj


class BaseSegment(metaclass=SegmentMetaclass):
    type = "base"


class RawSegment(BaseSegment):
    type = "raw"


print(BaseSegment.class_types)
print(RawSegment.class_types)
```

Reply:
🤯 learning new things every day! I've never used metaclasses before but I'll have a go.

Reply:
@barrywhart - ok I think I got the metaclasses working. Check out what I've done and see what you think.